feat(avm)!: Constrain tx `reverted` in public inputs

MirandaWood opened the pull request:
### Lookup `reverted` in public inputs

This PR constrains `reverted` in the tx trace against the public inputs. Previously, we just set `reverted` per phase and did not permeate it at the tx level. So this PR:

- Adds `tx_reverted` bool which is flipped from 0 to 1 if we revert at any phase
- Constrains that it starts matching `reverted` and is continuous per row, unless we have a reverted phase where it must be 1
- Looks up this value in the public inputs at the final (cleanup) stage

### Notes/TODOs

- Could likely combine some small relations into one if we prefer
- Could instead constrain we start with `tx_reverted = 0` if we cannot revert at the first row (`start_tx = 1`)

MirandaWood commented on barretenberg/cpp/pil/vm2/tx_context.pil:
```diff
@@ -168,6 +168,30 @@ namespace tx;
 
+    // Read reverted from PI
+
+    // Whether this tx reverted
+    pol commit tx_reverted;
+    #[INIT_TX_REVERTED]
+    start_tx * (reverted - tx_reverted) = 0;
+    // If reverted == 1, tx_reverted == 1, otherwise it must be continuous (see #[TX_REVERTED_CONTINUITY])
+    // from the start of the tx (see #[INIT_TX_REVERTED]).
+    #[SET_TX_REVERTED]
+    reverted * (1 - tx_reverted) = 0;
```
Could combine these if we prefer:

- If we cannot revert at `start_tx`, `tx_reverted` must start at 0, and could probably use the existing `#[TX_REVERTED_CONTINUITY]` to constrain init
- Both these relations set `tx_reverted = reverted`, so could be something like `(start_tx + reverted) * (reverted - tx_reverted) = 0`

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/tx_trace.cpp:
```diff
@@ -605,6 +608,9 @@ void TxTraceBuilder::process(const simulation::EventEmitterInterface<simulation:
         // We have events to process in this phase
         for (const auto& tx_phase_event : phase_events) {
+            // If this phase is the first revert, set tx_reverted:
+            tx_reverted |= tx_phase_event->reverted;
```
This works but I think it's formally speaking a bitwise OR. C++ doesn't have `||=` which is the boolean OR. To be safe it might be better to do `tx_reverted = tx_reverted || tx_phase_event->reverted;`

In this case it doesn't matter much but for example, `|=` does not do short=circuit and the above expression would.

MirandaWood replied:
Ah thanks, have changed!

dbanks12 reviewed (approved):
LGTM!

MirandaWood committed:
feat(avm)!: Constrain tx `reverted` in public inputs

### Lookup `reverted` in public inputs

This PR constrains `reverted` in the tx trace against the public inputs. Previously, we just set `reverted` per phase and did not permeate it at the tx level. So this PR:

- Adds `tx_reverted` bool which is flipped from 0 to 1 if we revert at any phase
- Constrains that it starts matching `reverted` and is continuous per row, unless we have a reverted phase where it must be 1
- Looks up this value in the public inputs at the final (cleanup) stage

### Notes/TODOs

- Could likely combine some small relations into one if we prefer
- Could instead constrain we start with `tx_reverted = 0` if we cannot revert at the first row (`start_tx = 1`)
